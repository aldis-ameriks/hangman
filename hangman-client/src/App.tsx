import { inject, observer } from 'mobx-react';
import React, { Component } from 'react';
import GameInput from './hangman/GameInput';
import GameState from './hangman/GameState';
import HangmanStore from './hangman/HangmanStore';
import NewGameButton from './hangman/NewGameButton';
import Typewriter from './hangman/Typewriter';

@inject('hangmanStore')
@observer
class App extends Component<{ hangmanStore: HangmanStore }> {
  public componentDidMount() {
    this.props.hangmanStore.initializeGame();
  }

  public componentDidUpdate() {
    this.renderGallows();
  }

  public render() {
    const { hangmanStore } = this.props;
    if (hangmanStore.gameState === 'loading') {
      return null;
    }

    const {
      notification,
      makeMove,
      startNewGame,
      isGameOver,
      letters,
      used,
      turnsLeft,
    } = hangmanStore;
    return (
      <>
        <div className="navbar navbar-inverse navbar-fixed-top">
          <div className="navbar-inner">
            <div className="container">
              <a className="brand" href="#">
                Hangman
              </a>
            </div>
          </div>
        </div>
        <div className="container">
          <Typewriter text="Welcome to Hangman" interval={100} render={text => <h1>{text}</h1>} />
        </div>
        <hr />
        <div className="container game-container">
          <div className="row">
            <div className="span6">
              <canvas id="gallows" height="504px" width="300px">
                Canvas is not supported ;-(
              </canvas>
            </div>
            <div className="span6">
              <GameState letters={letters} notification={notification} turnsLeft={turnsLeft} />
              {isGameOver ? (
                <NewGameButton startNewGame={startNewGame} />
              ) : (
                <GameInput makeMove={makeMove} used={used} />
              )}
            </div>
          </div>
        </div>
      </>
    );
  }

  private renderGallows() {
    const { turnsLeft } = this.props.hangmanStore;
    // @ts-ignore
    const canvas = document.getElementById('gallows');
    // @ts-ignore
    if (!canvas || !canvas.getContext('2d')) {
      return null;
    }

    // @ts-ignore
    const ctx = canvas.getContext('2d');
    ctx.strokeStyle = '#fff';
    ctx.lineWidth = 4;

    switch (turnsLeft) {
      case 7: {
        ctx.beginPath();
        ctx.clearRect(0, 0, 400, 504);
        ctx.stroke();

        ctx.moveTo(350, 450);
        ctx.lineTo(50, 450);
        ctx.lineTo(50, 50);
        ctx.lineTo(200, 50);
        ctx.stroke();
        break;
      }
      case 6: {
        ctx.beginPath();
        ctx.moveTo(200, 50);
        ctx.lineTo(200, 100);
        ctx.stroke();
        break;
      }
      case 5: {
        ctx.beginPath();
        ctx.moveTo(200, 100);
        ctx.arc(200, 150, 30, 1.5 * Math.PI, -0.5 * Math.PI, true);
        ctx.stroke();
        break;
      }
      case 4: {
        ctx.beginPath();
        ctx.moveTo(200, 180);
        ctx.lineTo(200, 300);
        ctx.stroke();
        break;
      }
      case 3: {
        ctx.beginPath();
        ctx.moveTo(200, 200);
        ctx.lineTo(150, 250);
        ctx.stroke();
        break;
      }
      case 2: {
        ctx.beginPath();
        ctx.moveTo(200, 200);
        ctx.lineTo(250, 250);
        ctx.stroke();
        break;
      }
      case 1: {
        ctx.beginPath();
        ctx.moveTo(200, 300);
        ctx.lineTo(150, 350);
        ctx.stroke();
        break;
      }
      case 0: {
        ctx.beginPath();
        ctx.moveTo(200, 300);
        ctx.lineTo(250, 350);

        ctx.moveTo(190, 140);
        ctx.lineTo(195, 145);
        ctx.moveTo(195, 140);
        ctx.lineTo(190, 145);

        ctx.moveTo(210, 140);
        ctx.lineTo(215, 145);
        ctx.moveTo(215, 140);
        ctx.lineTo(210, 145);
        ctx.stroke();

        ctx.beginPath();
        ctx.arc(200, 170, 14, -0.15 * Math.PI, -0.85 * Math.PI, true);
        ctx.stroke();

        break;
      }
    }
  }
}

export default App;
