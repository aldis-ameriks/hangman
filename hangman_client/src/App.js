import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import Grid from '@material-ui/core/Grid';

import { ReactComponent as Gallows } from './components/gallows.svg';
import HangmanProvider from './hangman/HangmanProvider';
import GameState from './hangman/GameState';
import GameInput from './hangman/GameInput';
import NewGameButton from './hangman/NewGameButton';
import NavigationBar from './components/NavigationBar';

const styles = theme => ({
  layout: {
    marginLeft: 'auto',
    marginRight: 'auto',
    marginTop: theme.spacing.unit * 4,
    maxWidth: 1000
  },
  paper: {
    padding: `${theme.spacing.unit * 2}px ${theme.spacing.unit * 3}px ${theme.spacing.unit * 3}px`,
    margin: `${theme.spacing.unit * 2}px`
  },
  title: {
    marginTop: theme.spacing.unit * 2
  },
  button: {
    marginTop: theme.spacing.unit * 4,
    display: 'block'
  }
});

const HangmanGame = ({ letters, notification, turnsLeft, used, isGameOver, classes, startNewGame, makeMove }) => (
  <Grid container spacing={40}>
    <Grid item sm={6} xs={12}>
      <Gallows />
    </Grid>
    <Grid item sm={6} xs={12}>
      <GameState letters={letters} notification={notification} turnsLeft={turnsLeft} used={used} />
      {isGameOver ? (
        <NewGameButton classes={classes} startNewGame={startNewGame} />
      ) : (
        <GameInput makeMove={makeMove} classes={classes} />
      )}
    </Grid>
  </Grid>
);

HangmanGame.propTypes = {
  letters: PropTypes.string.isRequired,
  used: PropTypes.string.isRequired,
  turnsLeft: PropTypes.number.isRequired,
  isGameOver: PropTypes.bool.isRequired,
  startNewGame: PropTypes.func.isRequired,
  makeMove: PropTypes.func.isRequired,
  classes: PropTypes.shape({}).isRequired,
  notification: PropTypes.shape({ type: PropTypes.string, message: PropTypes.string })
};

HangmanGame.defaultProps = {
  notification: null
};

const App = ({ classes }) => (
  <>
    <NavigationBar />
    <main className={classes.layout}>
      <Paper className={classes.paper}>
        <HangmanProvider render={gameState => <HangmanGame classes={classes} {...gameState} />} />
      </Paper>
    </main>
  </>
);

App.propTypes = {
  classes: PropTypes.shape({}).isRequired
};

export default withStyles(styles)(App);
