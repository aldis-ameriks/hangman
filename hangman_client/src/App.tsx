import Grid from '@material-ui/core/Grid';
import Paper from '@material-ui/core/Paper';
import { withStyles } from '@material-ui/core/styles';
import React, { Component } from 'react';

import CircularProgress from '@material-ui/core/CircularProgress/CircularProgress';
import { inject, observer } from 'mobx-react';
import { ReactComponent as Gallows } from './components/gallows.svg';
import NavigationBar from './components/NavigationBar';
import GameInput from './hangman/GameInput';
import GameState from './hangman/GameState';
import HangmanStore, { HangmanGameState } from './hangman/HangmanStore';
import NewGameButton from './hangman/NewGameButton';

const styles = (theme: any) => ({
  layout: {
    marginLeft: 'auto',
    marginRight: 'auto',
    marginTop: theme.spacing.unit * 4,
    maxWidth: 1000,
  },
  paper: {
    padding: `${theme.spacing.unit * 2}px ${theme.spacing.unit * 3}px ${theme.spacing.unit * 3}px`,
    margin: `${theme.spacing.unit * 2}px`,
  },
  title: {
    marginTop: theme.spacing.unit * 2,
  },
  button: {
    marginTop: theme.spacing.unit * 4,
    display: 'block',
  },
});

type HangmanGameProps = HangmanGameState & { classes: any };

const HangmanGame: React.FunctionComponent<HangmanGameProps> = ({
  letters,
  notification,
  turnsLeft,
  used,
  isGameOver,
  classes,
  startNewGame,
  makeMove,
}) => (
  <Grid container={true} spacing={40}>
    <Grid item={true} sm={6} xs={12}>
      <Gallows />
    </Grid>
    <Grid item={true} sm={6} xs={12}>
      <GameState letters={letters} notification={notification} turnsLeft={turnsLeft} used={used} />
      {isGameOver ? (
        <NewGameButton classes={classes} startNewGame={startNewGame} />
      ) : (
        <GameInput makeMove={makeMove} classes={classes} />
      )}
    </Grid>
  </Grid>
);

HangmanGame.defaultProps = {
  notification: undefined,
};

@inject('hangmanStore')
@observer
class App extends Component<{ classes: any; hangmanStore: HangmanStore }> {
  public componentDidMount() {
    this.props.hangmanStore.initializeGame();
  }

  public render() {
    const { classes, hangmanStore } = this.props;
    if (hangmanStore.gameState === 'loading') {
      return <CircularProgress style={{ display: 'block', margin: 'auto' }} />;
    }

    const {
      notification,
      makeMove,
      startNewGame,
      isGameOver,
      letters,
      used,
      turnsLeft,
    } = hangmanStore;
    return (
      <>
        <NavigationBar />
        <main className={classes.layout}>
          <Paper className={classes.paper}>
            <HangmanGame
              classes={classes}
              letters={letters}
              isGameOver={isGameOver}
              makeMove={makeMove}
              notification={notification}
              startNewGame={startNewGame}
              turnsLeft={turnsLeft}
              used={used}
            />
          </Paper>
        </main>
      </>
    );
  }
}

export default withStyles(styles)(App);
