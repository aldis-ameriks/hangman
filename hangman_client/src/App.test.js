import React from 'react';
import { render } from 'react-testing-library';
import App from './App';

describe('App', () => {
  it('initially', () => {
    // TODO: What's the best approach for mocking sockets?
    const component = render(<App />);
    expect(component.baseElement).toMatchSnapshot();
  });
});
