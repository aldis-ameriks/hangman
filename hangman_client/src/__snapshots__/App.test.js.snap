// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App initially renders loader 1`] = `
<body>
  <div>
    <header
      class="MuiPaper-root-14 MuiPaper-elevation4-20 MuiAppBar-root-5 MuiAppBar-positionStatic-9 MuiAppBar-colorDefault-11"
    >
      <div
        class="MuiToolbar-root-41 MuiToolbar-regular-43 MuiToolbar-gutters-42"
      >
        <h2
          class="MuiTypography-root-45 MuiTypography-title-51 MuiTypography-colorInherit-74"
        >
          Hangyman
        </h2>
      </div>
    </header>
    <main
      class="App-layout-1"
    >
      <div
        class="MuiPaper-root-14 MuiPaper-elevation2-18 MuiPaper-rounded-15 App-paper-2"
      >
        <div
          class="MuiCircularProgress-root-80 MuiCircularProgress-colorPrimary-83 MuiCircularProgress-indeterminate-82"
          role="progressbar"
          style="width: 40px; height: 40px; display: block; margin: auto;"
        >
          <svg
            class="MuiCircularProgress-svg-85"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle-86 MuiCircularProgress-circleIndeterminate-88"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </div>
      </div>
    </main>
  </div>
</body>
`;
